File: apple/Sources/ReactNativeWatchOS/Modifiers/ContentShape.swift
import SwiftUI

/// SwiftUI `.contentShape(_:)`. Sets the view's hit-testing region to a
/// simple shape. The shape branches are switched individually because each
/// concrete `Shape` type differs — a `RoundedRectangle` is not the same type
/// as a `Circle`. Defaults to `rectangle` when the param is missing/unknown.
enum RNWContentShapeModifier {
    @MainActor
    static func register(into r: RNWModifierRegistry) {
        r.register("contentShape") { view, params, _ in
            switch params.string("shape") {
            case "circle":
                return AnyView(view.contentShape(Circle()))
            case "capsule":
                return AnyView(view.contentShape(Capsule()))
            case "roundedRectangle":
                let radius = params.cgFloat("cornerRadius") ?? 0
                return AnyView(view.contentShape(
                    RoundedRectangle(cornerRadius: radius)
                ))
            default:
                return AnyView(view.contentShape(Rectangle()))
            }
        }
    }
}

File: src/modifiers/contentShape.ts
import { createModifier } from './createModifier';

export type ContentShapeKind =
  | 'rectangle'
  | 'circle'
  | 'capsule'
  | 'roundedRectangle';

export interface ContentShapeParams {
  /// The shape used for hit testing (and interaction/preview shaping).
  /// Defaults to `'rectangle'`.
  shape?: ContentShapeKind;
  /// Corner radius in points, used only when `shape` is `'roundedRectangle'`.
  cornerRadius?: number;
}

/// SwiftUI `.contentShape(_:)`. Defines the hit-testing area of the view
/// using a simple shape. Useful to make an entire (partly transparent)
/// region tappable, or to clip the touchable area to a circle/capsule.
export function contentShape(
  shape?: ContentShapeKind | ContentShapeParams
) {
  if (typeof shape === 'string') {
    return createModifier('contentShape', { shape });
  }
  return createModifier('contentShape', shape ?? {});
}
